libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1132 @@
+        {
+          *is_filtered = TRUE;
+          return TRUE;
+        }
+    }
+
+  *is_filtered = FALSE;
+  return TRUE;
+}
+
+static gboolean
+action_filter_applies (JsonObject   *object,
+                       const gchar  *filter_key_name,
+                       gboolean     *is_filtered,
+                       GError      **error)
+{
+  JsonNode *filter_value = json_object_get_member (object, filter_key_name);
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  if (g_strcmp0 (filter_key_name, "~architectures") == 0)
+    return architecture_skip_filter_applies (filter_value, is_filtered, error);
+
+  g_set_error (error,
+               EOS_UPDATER_ERROR,
+               EOS_UPDATER_ERROR_UNKNOWN_ENTRY_IN_AUTOINSTALL_SPEC,
+               "Unknown action filter value '%s', expected one of '~architectures'",
+               filter_key_name);
+  return FALSE;
+}
+
+/* We do this at the same time as reading the JSON node so that we don't have
+ * to keep filter information around in memory */
+static gboolean
+action_node_should_be_filtered_out (JsonNode  *node,
+                                    gboolean  *is_filtered,
+                                    GError   **error)
+{
+  JsonObject *object = json_node_get_object (node);
+  JsonNode *filters_object_node = json_object_get_member (object, "filters");
+  JsonObject *filters_object = NULL;
+  g_autoptr(GList) filters_object_keys = NULL;
+  GList *iter = NULL;
+
+  g_return_val_if_fail (is_filtered != NULL, FALSE);
+
+  /* No filters, so this action cannot be filtered out */
+  if (filters_object_node == NULL)
+    {
+      *is_filtered = FALSE;
+      return TRUE;
+    }
+
+  filters_object = json_node_get_object (filters_object_node);
+
+  if (filters_object == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected 'filters' node to be an object, but was %s", node_str);
+
+      return FALSE;
+    }
+
+  filters_object_keys = json_object_get_members (filters_object);
+
+  for (iter = filters_object_keys; iter != NULL; iter = iter->next)
+    {
+      gboolean action_is_filtered_on_this_filter;
+      g_autoptr(GError) local_error = NULL;
+
+      if (!action_filter_applies (filters_object,
+                                  iter->data,
+                                  &action_is_filtered_on_this_filter,
+                                  &local_error))
+        {
+          if (g_error_matches (local_error,
+                               EOS_UPDATER_ERROR,
+                               EOS_UPDATER_ERROR_UNKNOWN_ENTRY_IN_AUTOINSTALL_SPEC))
+            {
+              g_warning ("%s. Skipping this filter and not applying action, system may be in an inconsistent state from this point forward", local_error->message);
+              *is_filtered = TRUE;
```

Comment:
Not sure I would want to filter the node out in this case; if we found something that we don't know in the autoinstall spec, I would lean towards not applying any filter. Otherwise, you could simply bubble up the error and let the caller decide what to do.

Reply:
I suppose with an `EOS_UPDATER_ERROR_UNKNOWN_FILTER_IN_AUTOINSTALL_FILE` ?